### ISSUE TITLE
Make an encoder that does not require ownership of the underlying buffer

### ISSUE THREAD

**Issue by dlrobertson** [mozilla/neqo#2054]
See [neqo-transport/src/tparams.rs:690](https://github.com/mozilla/neqo/blob/165fff6c11f1c863ec0b2ae978dd572f65155383/neqo-transport/src/tparams.rs#L690):

```rust
    // TODO: Modify to avoid a copy.
    let mut enc = Encoder::default();
    self.local.encode(&mut enc);
    assert!(enc.len() <= d.len());
    d[..enc.len()].copy_from_slice(enc.as_ref());
```

It should be possible to create a codec that doesn't require ownership of the underlying buffer. My underlying long-term goal would be something akin to what we had for [smoltcp](https://github.com/smoltcp-rs/smoltcp), where the packet structure was able to abstract over mutability (e.g. [the ipv6 Packet structure](https://github.com/smoltcp-rs/smoltcp/blob/8e3ea5c7f09a76f0a4988fda20cadc74eacdc0d8/src/wire/ipv6.rs#L537)).

**Comment by dlrobertson** [comment]
This TODO has bugged me for a while and I have a long flight with a good bit of time in the airport today 😆 so I'll make a first pass at doing this... Comments and critiques on this are very much appreciated.

**Comment by mxinden** [comment]
:+1: we will need an `Encoder` encoding into an externally managed buffer (send buffer) as part of https://github.com/mozilla/neqo/issues/1693 as well.

**Comment by dlrobertson** [comment]
### TL;DR

I let my enthusiasm get the better of me 😆 The current usage of the `Encoder` structure doesn't really play well with a backing buffer that is not growable.

### Problem

There are some cases where the `Encoder`'s `capacity` is treated as a limit (e.g. the [neqo-thransport PacketBuilder](https://github.com/mozilla/neqo/blob/main/neqo-transport/src/packet/mod.rs#L121)). I think cases like this could play nice with a backing buffer of a fixed size with a managed offset. There are however many locations where this doesn't work so well, so I think it would be a significant lift to get a strict "[smoltcp]-like" packet structure.

### Possible solution

I think the easiest solution would be to introduce a enum for the buffer backing the encoder with a `Borrowed` variant for slices and a `Owned` variant for `Vec`, similar to [ManagedSlice](https://docs.rs/managed/latest/managed/enum.ManagedSlice.html). On encode, the `Borrowed` case could assert that the buffer still has space, and the Owned case could expand in size if needed. There may be other solutions, but I think this would be the simplest to implement. Comments and critiques would be very much welcomed.

[smoltcp]: https://github.com/smoltcp-rs/smoltcp
